portfolio/src/main/java/com/google/sps/servlets/VoteServlet.java

```diff
@@ -40,37 +43,71 @@
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String parsedBody = CharStreams.toString(request.getReader());
-
-    JsonParser parser = new JsonParser();
-    JsonElement parsedJson = parser.parse(parsedBody);
-    JsonObject jsonObject = parsedJson.getAsJsonObject();
+    JsonObject jsonVote = UtilityFunctions.stringToJsonObject(parsedBody);
 
     long commentId = Long.parseLong(UtilityFunctions.getFieldFromJsonObject(
-        jsonObject, "id", "0"));
-    if (commentId != 0) {
+        jsonVote, "id", "0"));
+    long amount = Long.parseLong(UtilityFunctions.getFieldFromJsonObject(
+        jsonVote, "amt", "0"));
+
+    if (commentId != 0 && amount != 0) {
       boolean isUpvote = Boolean.parseBoolean(UtilityFunctions.getFieldFromJsonObject(
-          jsonObject, "isupvote", "true"));
-      changeVoteInDatastore(commentId, isUpvote);
+        jsonVote, "isupvote", "true"));
+      changeVoteInDatastore(commentId, isUpvote, amount);
     }
   }
 
   /* 
-   * Changes the number of upvotes/downvotes of the comment IDed by commentId
-   * based on whether isUpvote is true or false as well as the net score of the comment
+   * Registers that the comment represented by commentId has been upvoted (if isUpvote
+   * is true) or downvoted (if isUpvote is false) if amount is 1 and the user hasn't
+   * already upvoted or downvoted the same comment. If they have, no change occurs.
+   * If amount is -1 and the user has upvoted [downvoted] a comment and isUpvote is true
+   * [false] then the vote is deregistered so that the user has no vote towards this comment.
+   * If amount is -1 and isUpvote is true and the user has downvoted the comment, no change 
+   * occurs.
    */
-  private void changeVoteInDatastore(long commentId, boolean isUpvote) {
+  private void changeVoteInDatastore(long commentId, boolean isUpvote, long amount) {
     Datastore datastore = DatastoreOptions.getDefaultInstance().getService();
     KeyFactory keyFactory = datastore.newKeyFactory().setKind("Comment");
     Entity comment = datastore.get(keyFactory.newKey(commentId));
+    
+    String voters = comment.getString("voters");
+    JsonObject obj = UtilityFunctions.stringToJsonObject(voters);
+    String userId = UtilityFunctions.getCurrentUserId();
+
+    /* 
+     * The user has upvoted a comment and is trying to downvote it 
+     * or has downvoted the comment and is trying to upvote it. In
+     * this case, no change should occur. 
+     */
```

Comment:
nit: move this inside the if-block because the comment applies to only the if-branch, not the entire if-elseif-else block.